### PR TITLE
Add conversion of CompletionStage to Deferred

### PR DESCRIPTION
**Pull request by jcornaz**
Hello,

This PR Implements `fun <T> CompletionStage<T>.asDeferred(): Deferred<T>` as proposed by @elizarov  in #262.

Let me know what you think about it.

### CODE REVIEW

**Comment by fvasco** [comment]
You should check `isDone`, invoking `whenComplete` on completed futures is expensive.

**Comment by jcornaz** [comment]
Yes indeed.

As `CompletableStage` doesn't provide any `isDone`  function I wrote:
`if (this is Future<*> && isDone()) return CompletableDeferred(get() as T)`

**Comment by fvasco** [comment]
Nice,
you should add the test for `CompletableFuture.failedFuture`.

Both `CompletableDeferred` and `CompletableFuture` have `completeExceptionally`, for clarity use `also` instead of `apply`.

**Comment by jcornaz** [comment]
Okay, I fixed it. I wasn't expecting the correct exception type in the test. A future created as failed wrap the cause exception in `ExecutionException` while a failure during the task is wrapped in `CompletionException`.

File: integration/kotlinx-coroutines-jdk8/src/test/kotlin/kotlinx/coroutines/experimental/future/FutureTest.kt [hunk 1]
```diff
@@ -255,6 +255,49 @@ class FutureTest : TestBase() {
...
+            fail("deferred.await() should throw an exception")
+        } catch (e: Exception) {
+            assertEquals("something went wrong", e.cause!!.message)
+        }
+    }
+
+    @Test
+    fun testCompletableFutureWithExceptionAsDeferred() = runBlocking {
+        val deferred: Deferred<Int> = CompletableFuture.supplyAsync {
+            Thread.sleep(100)
```
Comment (fvasco) [inline integration/kotlinx-coroutines-jdk8/src/test/kotlin/kotlinx/coroutines/experimental/future/FutureTest.kt]: This `sleep` is useful?

File: integration/kotlinx-coroutines-jdk8/src/test/kotlin/kotlinx/coroutines/experimental/future/FutureTest.kt [hunk 2]
```diff
@@ -255,6 +255,49 @@ class FutureTest : TestBase() {
         assertThat(future.get(), IsEqual("OK"))
     }
 
+    @Test
+    fun testCompletableFutureStageAsDeferred() = runBlocking {
+        val deferred: Deferred<Int> = CompletableFuture.supplyAsync { Thread.sleep(100) ; 42 }.asDeferred()
```
Comment (fvasco) [inline integration/kotlinx-coroutines-jdk8/src/test/kotlin/kotlinx/coroutines/experimental/future/FutureTest.kt]: A Lock is more reliable than a `sleep`.
You can
1. use a Lock to pause the Supplier
2. check Deferred as incomplete
3. release the Lock to resume the Supplier
4. verify Deferred result

**Comment by fvasco** [comment]
Hi @jcornaz,
all your commits refer to a single proposal, you should squash all them in a single one (using `git rebase` to rewrite your commit history).

**Comment by jcornaz** [comment]
Thanks @fvasco. I wasn't convinced either by the `sleep`. And using a lock is indeed way better.

I used kotlinx.coroutines `Mutex` is that okay ?

**Comment by fvasco** [comment]
Using a `Mutex` in a `Supplier` is a bit tricky, try yourself (like using a `delay` instead of a `sleep`).

**Comment by jcornaz** [comment]
Of course I have to use it inside of `runBlocking` is that what you mean by "tricky" ?

**Comment by jcornaz** [comment]
Yes it is more readable I agree. I've replace the `Mutex` by a Java `Lock`

**Comment by elizarov** [comment]
Thanks a lot. Merged. I've made a few changes though. The most important one is that in the fast-path of `CompletionStage.asDeferred` when we are using `future.get()` we should unwrap `ExecutionException` that `get` uses just like it is done in `await` implementation. I've updated the tests correspondingly to make a stricter checks that the appropriate exception types are preserved.

I've also implemented two-way cancellation support `CompletionStage.asDeferred` in a similar vein to `Deferred.asCompletionStage`. It is just one-line addition. As a result, when the resulting deferred is cancelled, it now cancels the source `CompletionStage` is that is possible (if it implements `Future` interface, that is).

I've also took a fresh look on other functions in this file and made changes there, too. I've used the same `this is Future<*>` trick in `CompletionStage.await` implementation, made it cancellable (what I was originally thinking -- I don't know) and deprecated a specialized version of `await`.

**Comment by jcornaz** [comment]
Cool, thank you.

**Comment by pacher** [comment]
> invoking `whenComplete` on completed futures is expensive

@fvasco Hi, do you have anything I can read about performance issues of `whenComplete`? I could not find it and quick looking at the code did not help. And `CompletionStage<T>.await()` uses it under the hood.
Is it expensive only for completed futures? Is it fine to use it for "awating" the result of not yet completed ones?
Is there any better alternative? Should I use `CompletableDeffered` and `invokeOnCompletion` instead? In my experience it was even slower, but it was long time ago and just some observations, not a real benchmark.

**Comment by fvasco** [comment]
Hi, @pacher,
`whenComplete` is more expensive than `get`, that's all.
I don't know your use case, but please consider avoiding conversions when possible.

**Pull request closed without merging**